app/src/main/java/org/simple/clinic/patient/PatientRepository.kt

```diff
@@ -124,7 +153,7 @@ class PatientRepository @Inject constructor(
 
             val newOrUpdatedPatients = payloads.map { it.toDatabaseModel(newStatus = DONE) }
 
-            database.patientDao().save(newOrUpdatedPatients)
+            savePatients(newOrUpdatedPatients).blockingAwait()
```

Comment:
Whoaaa. I'm *very* paranoid of using `blockingAwait()` calls.

1. It's possible to trigger a race condition. I cannot recall the situations, but I think it can be done by calling await() on the main thread or something. 

2. It feels very wrong to mix async and sync code together.

Considering that this `flatMap` is doing multiple things now, my favorite way of solving this problem is by splitting the stream into separate jobs. What do you think?

```kotlin
fun mergeWithLocalData(serverPayloads: List<PatientPayload>): Completable {
  val payloadsToSave = serverPayloads
      .toObservable()
      .filter { payload ->
        val localCopy = database.patientDao().getOne(payload.uuid)
        localCopy?.syncStatus.canBeOverriddenByServerCopy()
      }
      .toList()
      .cache()

  val saveAddresses = payloadsToSave
      .map { payloads -> payloads.map { it.address.toDatabaseModel() } }
      .flatMapCompletable { addresses ->
        Completable.fromAction {
          database.addressDao().save(addresses)
        }
      }

  val savePatients = payloadsToSave
      .map { payloads -> payloads.map { it.toDatabaseModel(newStatus = DONE) } }
      .flatMapCompletable { patients -> savePatients(patients) }

  val savePhoneNumbers = payloadsToSave
      .map { payloads ->
        payloads
            .filter { it.phoneNumbers != null }
            .flatMap { payload ->
              payload.phoneNumbers!!.map { it.toDatabaseModel(payload.uuid) }
            }
      }
      .filter { numbers -> numbers.isNotEmpty() }
      .flatMapCompletable { numbers ->
        Completable.fromAction {
          database.phoneNumberDao().save(numbers)
        }
      }

  return saveAddresses.andThen(savePatients).andThen(savePhoneNumbers)
}
```

The obvious drawback is that this function is very long now.

Reply:
Oh! I added it in there because I wanted to see results on the UI first, and I planned to clean up the chain later. Guess I should've added a task somewhere.

Will fix this.